### ISSUE TITLE
【paddle.cast】转换为 unit16 时的bug

### ISSUE THREAD

**Issue by wzzju** [PaddlePaddle/Paddle#34927]
关于bfloat16的转换问题，cast后的输出应该是1，但是这里的输出是16256；
环境 paddle 2.1.0，Mac
![image](https://user-images.githubusercontent.com/17102274/129527325-1178565c-9aeb-403a-98d4-caace3675707.png)

竞品都是1：
tf: 2.5.0 Tesla T4
![image](https://user-images.githubusercontent.com/17102274/129527372-6c424bce-5716-4f58-9558-7fac340cfc87.png)

pytorch: 1.8.1 Tesla T4
![image](https://user-images.githubusercontent.com/17102274/129527448-f510b876-d108-4883-a6f5-7f59104628b1.png)

============== 问题排查 ==============
BF16问题。见 python/paddle/fluid/framework.py 中 convert_np_dtype_to_dtype，由于numpy不支持bfloat16，代码中将np.uint16转为paddle.bfloat16
![image](https://user-images.githubusercontent.com/17102274/129527486-740157ae-d2e4-499a-867a-bf8edbc9e23a.png)

============== 解决方案 ===============

建议：可以参考tensorflow，add bfloat16 as numpy extensions, refer to：

tf bfloat16 numpy extension way： https://stackoverflow.com/questions/66354058/tensorflow-register-numpy-bfloat16-extension

tf bfloat16 numpy extension commit：   https://github.com/tensorflow/tensorflow/commit/24ffe9f729160a095a5cab8f5923920182803182

**Comment by lidanqing-vv** [comment]
Temp comments

convert_bf16_to_fp32(a.cast(X, "bfloat16"))

str::cout<<a<<std::endl;

string()

**Comment by lidanqing-vv** [comment]
This issue is pending by https://github.com/PaddlePaddle/Paddle/pull/35332

**Comment by lidanqing-vv** [comment]
np.uint16 is used as bfloat16 because np.uint16 was reserved datatype, for nothing else, that is why np.uint16 is temporarily used to represent bfloat16.
With https://github.com/PaddlePaddle/Paddle/pull/35332/files merged, then in `python/paddle/fluid/framework.py`, we just need to `import paddle_bfloat`, and then do
```
 elif dtype == np.bfloat16:
        return core.VarDesc.VarType.BF16
 elseif dtype == np.uint16:
       print("Reserved datatype")
       return ...
```
@Aganlengzi Hi so that bfloat16 PR is the only blocker for this issue.

**Comment by lidanqing-vv** [comment]
@Aganlengzi Rewrite current paddle_bfloat16 is too big and even hard to estimate, we are supporting the Paddle community with mkldnn FP32/INT8/BF16 affinity, we actually don't have resources to implement numpy bfloat. Aligned with @jakpiase @wozna

**Comment by Aganlengzi** [comment]
> np.uint16 is used as bfloat16 because np.uint16 was reserved datatype, for nothing else, that is why np.uint16 is temporarily used to represent bfloat16. With https://github.com/PaddlePaddle/Paddle/pull/35332/files merged, then in `python/paddle/fluid/framework.py`, we just need to `import paddle_bfloat`, and then do
> 
> ```
>  elif dtype == np.bfloat16:
>         return core.VarDesc.VarType.BF16
>  elseif dtype == np.uint16:
>        print("Reserved datatype")
>        return ...
> ```
> 
> @Aganlengzi Hi so that bfloat16 PR is the only blocker for this issue.

Yes I found the related PR https://github.com/PaddlePaddle/Paddle/pull/27325

**Comment by Aganlengzi** [comment]
> @Aganlengzi Rewrite current paddle_bfloat16 is too big and even hard to estimate, we are supporting the Paddle community with mkldnn FP32/INT8/BF16 affinity, we actually don't have resources to implement numpy bfloat. Aligned with @jakpiase @wozna

@lidanqing-intel I asked internal PM to evaluate the paddle_bfloat16 package, it is ok to add the package into python/requirements.txt since it is a separate third-party package for Paddle. But as I commented in #35332, you have to create a new PR for reviewing because some CI pipelines can not get results.

**Comment by Aganlengzi** [comment]
@lidanqing-intel hi, Danqing, bfloat16 PR has been merged. Please continue this issue ~

**Comment by yaomichael** [comment]
@Aganlengzi can you close this ticket if no further action needed?

**Comment by lidanqing-vv** [comment]
@yaomichael  @Aganlengzi  Hi, this issue is caused by the same reason as https://github.com/PaddlePaddle/Paddle/issues/39473. I was thinking after Jakub's bf16 core RP https://github.com/PaddlePaddle/Paddle/pull/40323 is merged, he can also test this issue case. But if you prefer you can close it.

**Comment by Aganlengzi** [comment]
> @Aganlengzi can you close this ticket if no further action needed?

> @yaomichael @Aganlengzi Hi, this issue is caused by the same reason as https://github.com/PaddlePaddle/Paddle/issues/39473. I was thinking after Jakub's bf16 core RP https://github.com/PaddlePaddle/Paddle/pull/40323 is merged, he can also test this issue case. But if you prefer you can close it.

Still need  further action as @lidanqing-intel said.

**Comment by jakpiase** [comment]
Hi, #40323 should fix this issue. It's goal is to stop using uint16 as a placeholder for bf16 and to allow creating uint16 tensors inside Paddle core. I will handle using BF16 in the same way as tensorflow did. The estimated time of that PR is June if everything will go smoothly, but I am not very experienced in pybind, so some problems may occur in the meantime
